### PR TITLE
Workbench hover to align and correctly apply delay setting

### PR DESCRIPTION
**Pull request by benibenj**
This PR refactors the usage of the hover service across various components. Instead of directly using the hover service, a new `WorkbenchHoverDelegate` is created and used. This delegate encapsulates the hover service and provides a consistent interface for components to use. This change simplifies the code and makes it easier to manage hover behavior across the workbench. Compared to the previous implementation, the delay will adapt when the configuration is changed without having to reload.

### CODE REVIEW

File: src/vs/workbench/browser/hover.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	get delay(): number {
+		return this._delay;
+	}
+
+	constructor(
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IHoverService private readonly hoverService: IHoverService,
+	) {
+		this._delay = this.configurationService.getValue<number>('workbench.hover.delay');
+		this.configurationService.onDidChangeConfiguration(e => {
```
Comment (bpasero) [inline src/vs/workbench/browser/hover.ts:24]: Disposable

File: src/vs/workbench/browser/hover.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IHoverDelegate, IHoverDelegateOptions, IHoverWidget } from 'vs/base/browser/ui/iconLabel/iconHoverDelegate';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { IHoverService } from 'vs/platform/hover/browser/hover';
+
+export class WorkbenchHoverDelegate implements IHoverDelegate {
+
+	readonly placement = 'mouse';
```
Comment (bpasero) [inline src/vs/workbench/browser/hover.ts:12]: Should this be something the user can override?

File: src/vs/workbench/browser/hover.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IHoverDelegate, IHoverDelegateOptions, IHoverWidget } from 'vs/base/browser/ui/iconLabel/iconHoverDelegate';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { IHoverService } from 'vs/platform/hover/browser/hover';
+
+export class WorkbenchHoverDelegate implements IHoverDelegate {
```
Comment (bpasero) [inline src/vs/workbench/browser/hover.ts:10]: Maybe to make this similar to `WorkbenchList` I would move this where the `IHoverService` is in `platform`

File: src/vs/workbench/browser/hover.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+			if (e.affectsConfiguration('workbench.hover.delay')) {
+				this._delay = this.configurationService.getValue<number>('workbench.hover.delay');
+			}
+		});
+	}
+
+	showHover(options: IHoverDelegateOptions): IHoverWidget | undefined {
+		return this.hoverService.showHover({
+			...options,
+			persistence: {
```
Comment (bpasero) [inline src/vs/workbench/browser/hover.ts:34]: I think I would still allow the `options` to override everything in case someone wants to not `hideOnHover`

**Review by bpasero — CHANGES_REQUESTED** [review]
Thanks for driving this.

I think there are a lot more places where the custom hover is installed where maybe we could think about more reuse. To give some examples:

https://github.com/microsoft/vscode/blob/ec850828b9d07b7280d150f665fb9ca723d9272c/src/vs/workbench/browser/parts/statusbar/statusbarPart.ts#L147-L153

https://github.com/microsoft/vscode/blob/ec850828b9d07b7280d150f665fb9ca723d9272c/src/vs/base/browser/ui/iconLabel/iconLabelHover.ts#L165

**Comment by benibenj** [comment]
Adopted it in further cases. I extended it to also be able to enable if the hover should instantly be shown after hovering over another element. Also the consumer can now add options to override the defaults.

File: src/vs/workbench/contrib/terminal/browser/terminalView.ts [hunk 5]
```diff
@@ -257,7 +257,7 @@ export class TerminalViewPane extends ViewPane {
 						return;
 					}
 				};
-				return new ActionViewItem(action, panelOnlySplitAction, { icon: true, label: false, keybinding: this._getKeybindingLabel(action) });
+				return new ActionViewItem(action, panelOnlySplitAction, { ...options, icon: true, label: false, keybinding: this._getKeybindingLabel(action) });
 			}
 			case TerminalCommandId.SwitchTerminal: {
 				return this._instantiationService.createInstance(SwitchTerminalActionViewItem, action);
```
Comment (bpasero) [inline src/vs/workbench/contrib/terminal/browser/terminalView.ts:263]: Do we need to pass on the `options` also to `SwitchTerminalActionViewItem` and `SingleTerminalTabActionViewItem`?

File: src/vs/workbench/contrib/terminal/browser/terminalEditor.ts [hunk 6]
```diff
@@ -203,18 +204,18 @@ export class TerminalEditor extends EditorPane {
...
 
-	override getActionViewItem(action: IAction): IActionViewItem | undefined {
+	override getActionViewItem(action: IAction, options: IBaseActionViewItemOptions): IActionViewItem | undefined {
 		switch (action.id) {
 			case TerminalCommandId.CreateTerminalEditor: {
 				if (action instanceof MenuItemAction) {
 					const location = { viewColumn: ACTIVE_GROUP };
 					const actions = getTerminalActionBarArgs(location, this._terminalProfileService.availableProfiles, this._getDefaultProfileName(), this._terminalProfileService.contributedProfiles, this._terminalService, this._dropdownMenu);
-					const button = this._instantiationService.createInstance(DropdownWithPrimaryActionViewItem, action, actions.dropdownAction, actions.dropdownMenuActions, actions.className, this._contextMenuService, {});
+					const button = this._instantiationService.createInstance(DropdownWithPrimaryActionViewItem, action, actions.dropdownAction, actions.dropdownMenuActions, actions.className, this._contextMenuService, { hoverDelegate: options.hoverDelegate });
```
Comment (bpasero) [inline src/vs/workbench/contrib/terminal/browser/terminalEditor.ts:213]: Why not passing over all `options`?

File: src/vs/workbench/contrib/scm/browser/util.ts [hunk 7]
```diff
@@ -161,11 +161,11 @@ class StatusBarActionViewItem extends ActionViewItem {
 }
 
 export function getActionViewItemProvider(instaService: IInstantiationService): IActionViewItemProvider {
-	return action => {
+	return (action, options) => {
 		if (action instanceof StatusBarAction) {
 			return new StatusBarActionViewItem(action);
```
Comment (bpasero) [inline src/vs/workbench/contrib/scm/browser/util.ts:166]: Should `options` be passed over here as well?

File: src/vs/workbench/contrib/scm/browser/scmViewPane.ts [hunk 8]
```diff
@@ -2492,12 +2492,12 @@ class SCMInputWidget {
 
 		// Toolbar
 		this.toolbar = instantiationService2.createInstance(SCMInputWidgetToolbar, this.toolbarContainer, {
-			actionViewItemProvider: action => {
+			actionViewItemProvider: (action, options) => {
 				if (action instanceof MenuItemAction && this.toolbar.dropdownActions.length > 1) {
 					return instantiationService.createInstance(DropdownWithPrimaryActionViewItem, action, this.toolbar.dropdownAction, this.toolbar.dropdownActions, '', this.contextMenuService, { actionRunner: this.toolbar.actionRunner });
```
Comment (bpasero) [inline src/vs/workbench/contrib/scm/browser/scmViewPane.ts:2497]: Should `options` be passed over here as well?

File: src/vs/workbench/contrib/debug/browser/debugViewlet.ts [hunk 9]
```diff
@@ -93,7 +94,7 @@ export class DebugViewPaneContainer extends ViewPaneContainer {
 		}
 	}
 
-	override getActionViewItem(action: IAction): IActionViewItem | undefined {
+	override getActionViewItem(action: IAction, options: IBaseActionViewItemOptions): IActionViewItem | undefined {
```
Comment (bpasero) [inline src/vs/workbench/contrib/debug/browser/debugViewlet.ts:97]: Should `options` be passed into the created action view items below as well?

File: src/vs/workbench/contrib/debug/browser/debugToolBar.ts [hunk 10]
```diff
@@ -89,7 +90,7 @@ export class DebugToolBar extends Themable implements IWorkbenchContribution {
 		this.activeActions = [];
 		this.actionBar = this._register(new ActionBar(actionBarContainer, {
 			orientation: ActionsOrientation.HORIZONTAL,
-			actionViewItemProvider: (action: IAction) => {
+			actionViewItemProvider: (action: IAction, options: IBaseActionViewItemOptions) => {
 				if (action.id === FOCUS_SESSION_ID) {
 					return this.instantiationService.createInstance(FocusSessionActionViewItem, action, undefined);
```
Comment (bpasero) [inline src/vs/workbench/contrib/debug/browser/debugToolBar.ts:95]: Should `options` be passed over here as well and to the action item below?

File: src/vs/workbench/contrib/codeEditor/browser/outline/documentSymbolsTree.ts [hunk 11]
```diff
@@ -113,29 +115,22 @@ export class DocumentSymbolGroupRenderer implements ITreeRenderer<OutlineGroup,
 	}
 }
 
-export class DocumentSymbolRenderer implements ITreeRenderer<OutlineElement, FuzzyScore, DocumentSymbolTemplate> {
+export class DocumentSymbolRenderer extends Disposable implements ITreeRenderer<OutlineElement, FuzzyScore, DocumentSymbolTemplate> {
```
Comment (bpasero) [inline src/vs/workbench/contrib/codeEditor/browser/outline/documentSymbolsTree.ts:118]: Why now `Disposable`?

File: src/vs/workbench/browser/workbench.ts [hunk 12]
```diff
@@ -152,6 +154,9 @@ export class Workbench extends Layout {
 				const dialogService = accessor.get(IDialogService);
 				const notificationService = accessor.get(INotificationService) as NotificationService;
 
+				// Default Hover Delegate
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.ts:157]: Would move this into `renderWorkbench`

File: src/vs/workbench/browser/parts/titlebar/titlebarPart.ts [hunk 13]
```diff
@@ -264,7 +241,7 @@ export class BrowserTitlebarPart extends Part implements ITitlebarPart {
 	private readonly editorToolbarMenuDisposables = this._register(new DisposableStore());
 	private readonly layoutToolbarMenuDisposables = this._register(new DisposableStore());
 
-	private readonly hoverDelegate = new TitlebarPartHoverDelegate(this.hoverService, this.configurationService);
+	private readonly hoverDelegate;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/titlebarPart.ts:244]: Missing type.

File: src/vs/workbench/browser/parts/titlebar/titlebarPart.ts [hunk 14]
```diff
@@ -552,7 +530,7 @@ export class BrowserTitlebarPart extends Part implements ITitlebarPart {
 		// --- Editor Actions
 		const activeEditorPane = this.editorGroupsContainer.activeGroup?.activeEditorPane;
 		if (activeEditorPane && activeEditorPane instanceof EditorPane) {
-			const result = activeEditorPane.getActionViewItem(action);
+			const result = activeEditorPane.getActionViewItem(action, { hoverDelegate: this.hoverDelegate });
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/titlebarPart.ts:533]: I had somehow expected the `options` being passed into the `actionViewItemProvider` method, didn't you make it required? Or only in other places.

File: src/vs/workbench/browser/parts/statusbar/statusbarPart.ts [hunk 15]
```diff
@@ -184,11 +151,20 @@ class StatusbarPart extends Part implements IStatusbarEntryContainer {
 		@IWorkbenchLayoutService layoutService: IWorkbenchLayoutService,
 		@IContextMenuService private contextMenuService: IContextMenuService,
 		@IContextKeyService private readonly contextKeyService: IContextKeyService,
-		@IHoverService private readonly hoverService: IHoverService,
-		@IConfigurationService private readonly configurationService: IConfigurationService
 	) {
 		super(id, { hasTitle: false }, themeService, storageService, layoutService);
 
+		this.hoverDelegate = instantiationService.createInstance(WorkbenchHoverDelegate, 'element', true);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/statusbar/statusbarPart.ts:157]: This leaks a disposable?

File: src/vs/workbench/browser/parts/statusbar/statusbarPart.ts [hunk 16]
```diff
@@ -184,11 +151,20 @@ class StatusbarPart extends Part implements IStatusbarEntryContainer {
...
+		this.hoverDelegate.setOptions((_, focus?: boolean) => {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/statusbar/statusbarPart.ts:158]: Maybe nicer if this was a callback I can pass in when constructing the hover delegate?

File: src/vs/workbench/browser/parts/editor/editorTabsControl.ts [hunk 17]
```diff
@@ -452,17 +456,7 @@ export abstract class EditorTabsControl extends Themable implements IEditorTabsC
...
-			showHover: options => {
-				return this.hoverService.showHover({
-					...options,
-					persistence: {
-						hideOnHover: true
-					}
-				});
-			}
-		};
+		return this.tabsHoverDelegate;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorTabsControl.ts:459]: What about the:

```
persistence: {
						hideOnHover: true
					}
```

File: src/vs/workbench/browser/parts/editor/editorTabsControl.ts [hunk 18]
```diff
@@ -162,6 +164,8 @@ export abstract class EditorTabsControl extends Themable implements IEditorTabsC
 
 		this.renderDropdownAsChildElement = false;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorTabsControl.ts:166]: I wonder if we should use instant hovers for tabs

File: src/vs/workbench/browser/parts/editor/breadcrumbsControl.ts [hunk 19]
```diff
@@ -230,17 +230,7 @@ export class BreadcrumbsControl {
...
-			showHover: (options: IHoverOptions) => {
-				return this.hoverService.showHover({
-					...options,
-					persistence: {
-						hideOnHover: true
-					}
-				});
-			}
-		};
+		this._hoverDelegate = getDefaultHoverDelegate('element');
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/breadcrumbsControl.ts:233]: What about `hideOnHover: true`

File: src/vs/base/browser/ui/button/button.ts [hunk 20]
```diff
@@ -240,10 +243,16 @@ export class Button extends Disposable implements IButton {
 			}
 		}
 
+		let title;
```
Comment (bpasero) [inline src/vs/base/browser/ui/button/button.ts:246]: Missing type

File: src/vs/base/browser/ui/button/button.ts [hunk 21]
```diff
@@ -240,10 +243,16 @@ export class Button extends Disposable implements IButton {
...
 		if (typeof this.options.title === 'string') {
-			this._element.title = this.options.title;
+			title = this.options.title;
 		} else if (this.options.title) {
-			this._element.title = renderStringAsPlaintext(value);
+			title = renderStringAsPlaintext(value);
+		}
+		if (!this._hover) {
+			this._hover = setupCustomHover(getDefaultHoverDelegate('mouse'), this._element, title);
```
Comment (bpasero) [inline src/vs/base/browser/ui/button/button.ts:253]: Disposal is missing, so this will leak?

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/browser/parts/editor/editorTabsControl.ts (hunk 17 above)
Comment (benibenj) [reply src/vs/workbench/browser/parts/editor/editorTabsControl.ts:459]: drfault behaviour

File: src/vs/workbench/browser/parts/editor/editorTabsControl.ts (hunk 18 above)
Comment (benibenj) [reply src/vs/workbench/browser/parts/editor/editorTabsControl.ts:166]: Not for the meantime.

File: src/vs/workbench/contrib/debug/browser/debugToolBar.ts (hunk 10 above)
Comment (benibenj) [reply src/vs/workbench/contrib/debug/browser/debugToolBar.ts:95]: This is a selectactionitem which does not support `IBaseActionViewItems`. This should be fine for the meantime.

File: src/vs/workbench/contrib/terminal/browser/terminalEditor.ts (hunk 6 above)
Comment (benibenj) [reply src/vs/workbench/contrib/terminal/browser/terminalEditor.ts:213]: not supported by `DropdownWithPrimaryActionViewItem` options

File: src/vs/workbench/contrib/timeline/browser/timelinePane.ts [hunk 22]
```diff
@@ -1147,14 +1147,11 @@ class TimelineTreeRenderer implements ITreeRenderer<TreeElement, FuzzyScore, Tim
...
-		@IConfigurationService private readonly configurationService: IConfigurationService,
 	) {
+		super();
+
 		this.actionViewItemProvider = createActionViewItem.bind(undefined, this.instantiationService);
-		this._hoverDelegate = {
-			showHover: (options: IHoverDelegateOptions) => this.hoverService.showHover(options),
-			delay: <number>this.configurationService.getValue('workbench.hover.delay')
-		};
+		this._hoverDelegate = getDefaultHoverDelegate('element');
```
[the remark this answers is not in the document]
Comment (benibenj) [reply src/vs/workbench/contrib/timeline/browser/timelinePane.ts:1154]: Lists do not have instant hover currently so I don't think it should be the case here. Let's start off without it and we can easily change later.

File: src/vs/workbench/contrib/timeline/browser/timelinePane.ts [hunk 23]
```diff
@@ -1133,7 +1133,7 @@ export class TimelineListVirtualDelegate implements IListVirtualDelegate<TreeEle
 	}
 }
 
-class TimelineTreeRenderer implements ITreeRenderer<TreeElement, FuzzyScore, TimelineElementTemplate> {
+class TimelineTreeRenderer extends Disposable implements ITreeRenderer<TreeElement, FuzzyScore, TimelineElementTemplate> {
```
[the remark this answers is not in the document]
Comment (benibenj) [reply src/vs/workbench/contrib/timeline/browser/timelinePane.ts:1136]: not sure why it as added here, removed it

File: src/vs/workbench/contrib/terminal/browser/terminalView.ts (hunk 5 above)
Comment (benibenj) [reply src/vs/workbench/contrib/terminal/browser/terminalView.ts:263]: not supported as they are ?SelectActionViewItem`s

File: src/vs/workbench/contrib/codeEditor/browser/outline/documentSymbolsTree.ts (hunk 11 above)
Comment (benibenj) [reply src/vs/workbench/contrib/codeEditor/browser/outline/documentSymbolsTree.ts:118]: not cleaned up after new changes

File: src/vs/base/browser/ui/countBadge/countBadge.ts [hunk 24]
```diff
@@ -57,7 +60,13 @@ export class CountBadge {
 
 	private render() {
 		this.element.textContent = format(this.countFormat, this.count);
-		this.element.title = format(this.titleFormat, this.count);
+
+		const title = format(this.titleFormat, this.count);
+		if (!this.hover) {
+			this.hover = setupCustomHover(getDefaultHoverDelegate('mouse'), this.element, title);
```
[the remark this answers is not in the document]
Comment (benibenj) [reply src/vs/base/browser/ui/countBadge/countBadge.ts:66]: Reverting as this is a bit more involved. To be done in debt week

**Review by bpasero — APPROVED** [review]
I have feedback, but will file a new issue.

**Pull request merged**